src/wasm/cpp/tests/wasm-cpp-test.ts

```diff
@@ -11,56 +11,220 @@ import {assert} from '../../../platform/chai-web.js';
+    assert.deepStrictEqual(slotComposer.received, [
+      ['RenderTest', 'root', {template: 'abc', model: {foo: 'bar'}}],
+      ['RenderTest', 'root', {model: {foo: 'bar'}}],
+      ['RenderTest', 'root', {template: 'abc'}],
+      ['RenderTest', 'root', {template: 'abc', model: {foo: 'bar'}}]
+    ]);
+  });
+
+  it('autoRender', async () => {
+    const {arc, stores, slotComposer} = await setup(`
+      import '${schemasFile}'
+
+      resource DataResource
+        start
+        [{"txt": "initial"}]
+      store DataStore of Data in DataResource
+
+      particle AutoRenderTest in '${wasmFile}'
+        consume root
+        in Data data
+
+      recipe
+        copy DataStore as h1
+        slot 'rootslotid-root' as slot1
+        AutoRenderTest
+          consume root as slot1
+          data <- h1
+      `);
+    const data = stores.get('data') as VolatileSingleton;
+
+    await data.set({id: 'i1', rawData: {txt: 'update'}});
     await arc.idle;
-    const some = await output.get();
-    assert.notEqual(some.id, '', 'output entity should have an id');
-    assert.notEqual(some.id, 'i1', 'output id should not be the same as the input');
-    assert.deepEqual(some.rawData, {txt: 'abc!', flg: true});
+
+    // First renderSlot call is initiated by the runtime, before handles are synced.
+    // With auto-render enabled, the second call occurs after sync and the third on handle update.
+    assert.deepStrictEqual(slotComposer.received, [
+      ['AutoRenderTest', 'root', {template: 'empty', model: {}}],
+      ['AutoRenderTest', 'root', {template: 'initial', model: {}}],
+      ['AutoRenderTest', 'root', {template: 'update', model: {}}],
+    ]);
+  });
+
+  it('fireEvent', async () => {
+    const {arc, stores, slotComposer} = await setup(`
+      import '${schemasFile}'
+
+      particle EventsTest in '${wasmFile}'
+        consume root
+        out Data output
+
+      recipe
+        slot 'rootslotid-root' as slot1
+        EventsTest
+          consume root as slot1
+          output -> h1
+      `);
+    const output = stores.get('output') as VolatileSingleton;
+
+    const particle = slotComposer.consumers[0].consumeConn.particle;
+    arc.pec.sendEvent(particle, 'root', {handler: 'icanhazclick', data: 'ignored'});
+    await arc.idle;
+
+    assert.deepStrictEqual((await output.get()).rawData, {txt: 'event:root:icanhazclick'});
+  });
+
+  it('serviceRequest / serviceResponse / resolveUrl', async () => {
+    const {stores} = await setup(`
+      import '${schemasFile}'
+
+      particle ServicesTest in '${wasmFile}'
+        out [ServiceResponse] output
+
+      recipe
+        ServicesTest
+          output -> h1
+      `);
+    const output = stores.get('output') as VolatileCollection;
+
+    const results = (await output.toList()).map(e => e.rawData);
+    assert.lengthOf(results, 4);
+
+    assert.deepStrictEqual(results[0], {call: 'resolveUrl', payload: 'RESOLVED($resolve-me)'});
+
+    const verifyRandom = (result, tag) => {
```

Comment:
Can you add types on result and tag?

Reply:
I would really prefer not to. This is used on literally the next two lines and nowhere else. Adding types here would just be noise.

Reply:
The dream is to get to 100% type coverage so that we can switch on noImplicitAny. Either you add them now, or someone else will have to come along later and add them as part of a code health cleanup (or, more likely, we never get there, and implicit anys will forever plague us...)

Reply:
Good point; I've changed the structure to a loop to avoid excessive type cost instead :-/

Reply:
I think I prefer this new version anyway :-)